File: task2/Python/task2.py
import sys

def is_valid_local_part(local: str) -> bool:
    # Length check
    if not (6 <= len(local) <= 30):
        return False
    # Cannot start or end with a dot
    if local[0] == '.' or local[-1] == '.':
        return False
    # No consecutive dots
    if '..' in local:
        return False
    # Allowed characters: a-z, 0-9, ., +
    return all(c.isalnum() or c in '.+' for c in local) and local.islower()

def normalize(email: str) -> str:
    try:
        at_pos = email.index('@')
    except ValueError:
        return ""
    if at_pos == 0 or at_pos == len(email) - 1:
        return ""

    local = email[:at_pos]
    domain = email[at_pos:]

    if not is_valid_local_part(local):
        return ""

    new_local = []
    for c in local:
        if c == '+':
            break
        if c != '.':
            new_local.append(c)
    return ''.join(new_local) + domain

def main():
    print("1 - input from console\n2 - input from file")
    try:
        choice = int(input())
    except ValueError:
        print("Invalid choice")
        return

    emails = set()

    if choice == 1:
        n = int(input("Enter number of emails: "))
        for _ in range(n):
            email = input().strip()
            norm = normalize(email)
            if norm:
                emails.add(norm)
    elif choice == 2:
        filename = input("Enter file name: ")
        try:
            with open(filename, 'r') as f:
                for line in f:
                    for email in line.split():
                        norm = normalize(email)
                        if norm:
                            emails.add(norm)
        except IOError:
            print("File open error")
            return
    else:
        print("Invalid choice")
        return

    print("Number of unique emails:", len(emails))

if __name__ == "__main__":
    main()

File: task2/Go/task2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func isValidLocalPart(local string) bool {
	// Length check
	if len(local) < 6 || len(local) > 30 {
		return false
	}
	// Cannot start or end with a dot
	if local[0] == '.' || local[len(local)-1] == '.' {
		return false
	}
	// No consecutive dots
	if strings.Contains(local, "..") {
		return false
	}
	// Allowed characters: a-z, 0-9, ., +
	for _, c := range local {
		if !((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '.' || c == '+') {
			return false
		}
	}
	return true
}

func normalize(email string) string {
	atPos := strings.Index(email, "@")
	if atPos <= 0 || atPos == len(email)-1 {
		return ""
	}
	local := email[:atPos]
	domain := email[atPos:]

	if !isValidLocalPart(local) {
		return ""
	}

	var newLocal strings.Builder
	for _, c := range local {
		if c == '+' {
			break
		}
		if c != '.' {
			newLocal.WriteRune(c)
		}
	}
	return newLocal.String() + domain
}

func main() {
	var choice int
	fmt.Println("1 - input from console\n2 - input from file")
	fmt.Scan(&choice)

	emails := make(map[string]bool)

	if choice == 1 {
		var n int
		fmt.Print("Enter number of emails: ")
		fmt.Scan(&n)

		for i := 0; i < n; i++ {
			var email string
			fmt.Scan(&email)
			if norm := normalize(email); norm != "" {
				emails[norm] = true
			}
		}
	} else if choice == 2 {
		var filename string
		fmt.Print("Enter file name: ")
		fmt.Scan(&filename)

		file, err := os.Open(filename)
		if err != nil {
			fmt.Println("File open error")
			return
		}
		defer file.Close()

		scanner := bufio.NewScanner(file)
		scanner.Split(bufio.ScanWords)
		for scanner.Scan() {
			email := scanner.Text()
			if norm := normalize(email); norm != "" {
				emails[norm] = true
			}
		}
		if err := scanner.Err(); err != nil {
			fmt.Println("Error reading file")
			return
		}
	} else {
		fmt.Println("Invalid choice")
		return
	}

	fmt.Println("Number of unique emails:", len(emails))
}
